File: aurora-web/src/state/backtest.rs
// Copyright 2025 blingbling21
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! 回测任务状态管理

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// 回测任务状态
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum BacktestStatus {
    /// 等待执行
    Pending,
    /// 正在运行
    Running,
    /// 已完成
    Completed,
    /// 执行失败
    Failed,
}

/// 回测任务信息
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BacktestTask {
    /// 任务ID
    pub id: Uuid,
    /// 任务名称
    pub name: String,
    /// 配置文件路径
    pub config_path: String,
    /// 数据文件路径
    pub data_path: String,
    /// 任务状态
    pub status: BacktestStatus,
    /// 创建时间
    pub created_at: DateTime<Utc>,
    /// 开始时间
    pub started_at: Option<DateTime<Utc>>,
    /// 完成时间
    pub completed_at: Option<DateTime<Utc>>,
    /// 进度(0-100)
    pub progress: u8,
    /// 错误信息
    pub error: Option<String>,
    /// 回测结果(JSON格式)
    pub result: Option<serde_json::Value>,
}

impl BacktestTask {
    /// 创建新任务
    pub fn new(name: String, config_path: String, data_path: String) -> Self {
        Self {
            id: Uuid::new_v4(),
            name,
            config_path,
            data_path,
            status: BacktestStatus::Pending,
            created_at: Utc::now(),
            started_at: None,
            completed_at: None,
            progress: 0,
            error: None,
            result: None,
        }
    }

    /// 标记任务开始
    pub fn start(&mut self) {
        self.status = BacktestStatus::Running;
        self.started_at = Some(Utc::now());
    }

    /// 更新进度
    pub fn update_progress(&mut self, progress: u8) {
        self.progress = progress.min(100);
    }

    /// 标记任务完成
    pub fn complete(&mut self, result: serde_json::Value) {
        self.status = BacktestStatus::Completed;
        self.completed_at = Some(Utc::now());
        self.progress = 100;
        self.result = Some(result);
    }

    /// 标记任务失败
    pub fn fail(&mut self, error: String) {
        self.status = BacktestStatus::Failed;
        self.completed_at = Some(Utc::now());
        self.error = Some(error);
    }
}

/// 回测任务摘要信息（用于列表展示，不包含完整的回测结果）
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BacktestTaskSummary {
    /// 任务ID
    pub id: Uuid,
    /// 任务名称
    pub name: String,
    /// 配置文件路径
    pub config_path: String,
    /// 数据文件路径
    pub data_path: String,
    /// 任务状态
    pub status: BacktestStatus,
    /// 创建时间
    pub created_at: DateTime<Utc>,
    /// 开始时间
    pub started_at: Option<DateTime<Utc>>,
    /// 完成时间
    pub completed_at: Option<DateTime<Utc>>,
    /// 进度(0-100)
    pub progress: u8,
    /// 错误信息
    pub error: Option<String>,
}

impl From<&BacktestTask> for BacktestTaskSummary {
    fn from(task: &BacktestTask) -> Self {
        Self {
            id: task.id,
            name: task.name.clone(),
            config_path: task.config_path.clone(),
            data_path: task.data_path.clone(),
            status: task.status.clone(),
            created_at: task.created_at,
            started_at: task.started_at,
            completed_at: task.completed_at,
            progress: task.progress,
            error: task.error.clone(),
        }
    }
}

File: aurora-front/src/lib/api/backtest.ts
// Copyright 2025 blingbling21
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * 回测管理 API 服务
 */

import { get, post, del } from './client';
import type {
  ApiResponse,
  BacktestTask,
  BacktestTaskSummary,
  StartBacktestRequest,
  FullBacktestResult,
} from '@/types/api';

/**
 * 回测管理服务类
 */
export class BacktestService {
  /**
   * 获取所有回测任务列表（历史记录）
   * 返回任务摘要列表，不包含完整的回测结果数据
   */
  static async list(): Promise<ApiResponse<BacktestTaskSummary[]>> {
    return get<BacktestTaskSummary[]>('/backtest/history');
  }

  /**
   * 获取指定回测任务详情
   * 
   * @param taskId 任务 ID
   */
  static async get(taskId: string): Promise<ApiResponse<BacktestTask>> {
    return get<BacktestTask>(`/backtest/${encodeURIComponent(taskId)}`);
  }

  /**
   * 创建并启动回测任务
   * 
   * @param request 启动回测请求
   */
  static async start(
    request: StartBacktestRequest
  ): Promise<ApiResponse<{ task_id: string }>> {
    return post<{ task_id: string }>('/backtest/start', request);
  }

  /**
   * 删除回测任务
   * 
   * @param taskId 任务 ID
   */
  static async delete(taskId: string): Promise<ApiResponse<void>> {
    return del<void>(`/backtest/${encodeURIComponent(taskId)}`);
  }

  /**
   * 获取回测结果
   * 
   * @param taskId 任务 ID
   */
  static async getResult(
    taskId: string
  ): Promise<ApiResponse<FullBacktestResult>> {
    return get<FullBacktestResult>(
      `/backtest/result/${encodeURIComponent(taskId)}`
    );
  }

  /**
   * 创建 WebSocket 连接 URL
   * 
   * @param taskId 任务 ID
   */
  static getWebSocketUrl(taskId: string): string {
    // 从环境变量获取API基础URL
    const apiBaseUrl = process.env.NEXT_PUBLIC_API_BASE_URL || 'http://localhost:8080/api';
    
    // 提取主机地址（移除/api路径）
    const url = new URL(apiBaseUrl);
    const protocol = url.protocol === 'https:' ? 'wss:' : 'ws:';
    const host = url.host; // 这里会包含端口号
    
    // 构建WebSocket URL
    const wsUrl = `${protocol}//${host}/ws/backtest/${encodeURIComponent(taskId)}`;
    
    console.log('WebSocket URL:', wsUrl); // 调试日志
    return wsUrl;
  }
}

// 导出单例方法供直接使用
export const backtestApi = {
  list: () => BacktestService.list(),
  get: (taskId: string) => BacktestService.get(taskId),
  start: (request: StartBacktestRequest) => BacktestService.start(request),
  delete: (taskId: string) => BacktestService.delete(taskId),
  getResult: (taskId: string) => BacktestService.getResult(taskId),
  getWebSocketUrl: (taskId: string) => BacktestService.getWebSocketUrl(taskId),
};
